client/verta/verta/_internal_utils/model_dependencies.py

```diff
@@ -0,0 +1,98 @@
+    """List all the functions present in the provided class object."""
+    return inspect.getmembers(model_class, predicate=inspect.isfunction)
+
+
+def unwrap(func: Callable) -> Callable:
+    """Unwraps a function to get the original function object in a non-blocking manner."""
+    try:
+        return inspect.unwrap(func)  # returns last object in chain
+    except ValueError:  # if a cycle is encountered
+        return func
+
+
+def list_modules_in_function_definition(func: Callable) -> List[ModuleType]:
+    """List all modules called within the body of the provided function."""
+    _func = unwrap(func)
+    return [
+        value for key, value in inspect.getclosurevars(_func).globals.items()
+        if isinstance(value, ModuleType)
+    ]
+
+
+def list_modules_in_function_signature(func: Callable) -> List[ModuleType]:
+    """List all modules used in type hints in the provided function's arguments
+    and return type hint."""
+    _func = unwrap(func)
+    hints = get_type_hints(_func)
+    arg_hints = { k: v for k, v in hints.items() if k != 'return' }
+    return_hint = { k: v for k, v in hints.items() if k == 'return' }
+
+    modules = [ inspect.getmodule(value) for key, value in arg_hints.items() ]
+
+    if return_hint:
+        mod = inspect.getmodule(return_hint['return'])
+        if mod.__name__ == 'typing':
+            ret_ann = inspect.signature(_func).return_annotation
+            nested_args = ret_ann.__args__
+            for a in nested_args:
+                if inspect.isclass(a):
+                    modules.append(inspect.getmodule(a))
+        else:
+            modules.append(inspect.getmodule(return_hint['return']))
+    return modules
+
+
+def list_function_local_annotation_module_names(func: Callable) -> List[str]:
+    """Return a list of modules extracted from type annotations on function
+    locals.
+
+    Variable annotations within a function body do not have a lifetime beyond
+    the local namespace. This function extracts the annotations from the
+    function body when possible.
+
+    """
+    class AnnotationsCollector(ast.NodeVisitor):
+        """Collects AnnAssign nodes for 'simple' annotation assignments"""
+
+        def __init__(self):
+            self.annotations = {}
+
+        def visit_AnnAssign(self, node) -> None:
+            if node.simple:
+                # 'simple' == a single name, not an attribute or subscription, thus
+                # `node.target.id` should exist.
+                self.annotations[node.target.id] = node.annotation
+
+    source = textwrap.dedent(inspect.getsource(func))
+    mod = ast.parse(source)
+    assert mod.body and isinstance(mod.body[0], (ast.FunctionDef, ast.AsyncFunctionDef))
+    collector = AnnotationsCollector()
+    collector.visit(mod.body[0])
+    return [
+        ast.get_source_segment(source, node)
+        for node in collector.annotations.values()
+    ]
```

Comment:
This function appears to be returning the verbatim type annotation (e.g. `"spacy.Errors"`—an object) rather than the module name.

Reply:
And consequently `get_source_segment()` means that if—for whatever bizarre reason—they're using some other identifier

```python
sp_err = spacy.Errors
spacy_error: sp_err = self.make_spacy_error()
```

it'll return `sp_err`.

Reply:
Because of the way Python handles annotations on function local variables, they can't be accessed easily via `inspect` the way annotations in signatures can for example, hence this additional hackery to TRY and get them.    This is a trade-off we might just have to accept and document as a weakness for now.

Reply:
Also captured as future improvement